moquette-0.17/broker/src/main/java/io/moquette/broker/unsafequeues/QueuePool.java

```diff
@@ -0,0 +1,490 @@
+    }
+
+    private List<SegmentRef> recreateRecycledSegmentsBetween(SegmentRef toSegment) {
+        return recreateRecycledSegmentsBetween(null, toSegment);
+    }
+
+    private List<SegmentRef> recreateRecycledSegmentsBetween(SegmentRef fromSegment, SegmentRef toSegment) {
+        final List<SegmentRef> recreatedSegments = new LinkedList<>();
+        int prevPageId = 0;
+        if (fromSegment != null) {
+            prevPageId = fromSegment.pageId;
+            // holes after previous segment, to complete the page
+            recreatedSegments.addAll(recreateRecycledSegments(fromSegment.offset + segmentSize, allocator.getPageSize(), fromSegment.pageId));
+            prevPageId++;
+        }
+
+        // all the intermediate pages
+        for (; prevPageId < toSegment.pageId; prevPageId++) {
+            recreatedSegments.addAll(recreateRecycledSegments(0, allocator.getPageSize(), prevPageId));
+        }
+
+        // holes before the current segment
+        recreatedSegments.addAll(recreateRecycledSegments(0, toSegment.offset, toSegment.pageId));
+        return recreatedSegments;
+    }
+
+    private List<SegmentRef> recreateRecycledSegments(int fromOffset, int toOffset, int pageId) {
+        final List<SegmentRef> recreatedSegments = new LinkedList<>();
+        while (fromOffset != toOffset) {
+            recreatedSegments.add(new SegmentRef(pageId, fromOffset));
+            fromOffset = fromOffset + segmentSize;
+        }
+        return recreatedSegments;
+    }
+
+    private LinkedList<SegmentRef> decodeSegments(String s) {
+        final String[] segments = s.substring(s.indexOf("(") + 1, s.lastIndexOf(")"))
+                .split("\\), \\(");
+
+        LinkedList<SegmentRef> acc = new LinkedList<>();
+        for (String segment : segments) {
+            final String[] split = segment.split(",");
+            final int idPage = Integer.parseInt(split[0].trim());
+            final int offset = Integer.parseInt(split[1].trim());
+
+            acc.offer(new SegmentRef(idPage, offset));
+        }
+        return acc;
+    }
+
+    public Queue getOrCreate(String queueName) throws QueueException {
+        final QueueName queueN = new QueueName(queueName);
+        if (queues.containsKey(queueN)) {
+            return queues.get(queueN);
+        } else {
+            // create new queue with first empty segment
+            final Segment segment = nextFreeSegment();
+            //notify segment creation for queue in queue pool
+            segmentedCreated(queueName, segment);
+
+            // When a segment is freshly created the head must the last occupied byte,
+            // so can't be the start of a segment, but one position before, or in case
+            // of a new page, -1
+            final Queue queue = new Queue(queueName, segment, VirtualPointer.buildUntouched(), segment, VirtualPointer.buildUntouched(),
+                this.allocator, callback, this);
+            queues.put(queueN, queue);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **369 and 382**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and returns that reference to its caller at line: **383**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.